### ISSUE TITLE
Extraction for Static Array Initializers

### ISSUE THREAD

**Issue by strongrex2001** [djkaty/Il2CppInspector#59]
For example, if I write
```csharp
using System;
public class C {
    static byte[] a = new byte[] { 1, 2, 3, 1, 2, 3, 1, 2, 3 };
}
```
The IL generated is
```asm
.class private auto ansi '<Module>'
{
} // end of class <Module>

.class public auto ansi beforefieldinit C
    extends [System.Private.CoreLib]System.Object
{
    // Fields
    .field private static uint8[] a

    // Methods
    .method public hidebysig specialname rtspecialname 
        instance void .ctor () cil managed 
    {
        // Method begins at RVA 0x2050
        // Code size 7 (0x7)
        .maxstack 8

        IL_0000: ldarg.0
        IL_0001: call instance void [System.Private.CoreLib]System.Object::.ctor()
        IL_0006: ret
    } // end of method C::.ctor

    .method private hidebysig specialname rtspecialname static 
        void .cctor () cil managed 
    {
        // Method begins at RVA 0x2058
        // Code size 24 (0x18)
        .maxstack 8

        IL_0000: ldc.i4.s 9
        IL_0002: newarr [System.Private.CoreLib]System.Byte
        IL_0007: dup
        IL_0008: ldtoken field valuetype '<PrivateImplementationDetails>'/'__StaticArrayInitTypeSize=9' '<PrivateImplementationDetails>'::CA05F63840607885D76B640E53F99D2087CB6116
        IL_000d: call void [System.Private.CoreLib]System.Runtime.CompilerServices.RuntimeHelpers::InitializeArray(class [System.Private.CoreLib]System.Array, valuetype [System.Private.CoreLib]System.RuntimeFieldHandle)
        IL_0012: stsfld uint8[] C::a
        IL_0017: ret
    } // end of method C::.cctor

} // end of class C

.class private auto ansi sealed '<PrivateImplementationDetails>'
    extends [System.Private.CoreLib]System.Object
{
    .custom instance void [System.Private.CoreLib]System.Runtime.CompilerServices.CompilerGeneratedAttribute::.ctor() = (
        01 00 00 00
    )
    // Nested Types
    .class nested private explicit ansi sealed '__StaticArrayInitTypeSize=9'
        extends [System.Private.CoreLib]System.ValueType
    {
        .pack 1
        .size 9

    } // end of class __StaticArrayInitTypeSize=9


    // Fields
    .field assembly static initonly valuetype '<PrivateImplementationDetails>'/'__StaticArrayInitTypeSize=9' CA05F63840607885D76B640E53F99D2087CB6116 at I_000027C4
    .data cil I_000027C4 = bytearray (
        01 02 03 01 02 03 01 02 03
    )

} // end of class <PrivateImplementationDetails>
```
It is stored by means of `.data cil`, and may be important for static analysis. It is nice to have them extracted.
(Is it in metadata or in code?)

**Comment by strongrex2001** [comment]
(This comment is mainly a question)
If I got it right, in `libil2cpp.so` (armv7), I could only find `dword_some_number` in the call to `RuntimeHelpers_InitializeArray`, which is a variable in section `.bss` with no writing xref to it. I was able to attach a debugger to the target application to complete the task, but it seems strange for me. Why is that?

**Comment by djkaty** [comment]
Without being able to look at the binary it's hard to guess what might be happening, however if an array's default content is all-zero then it won't be mapped in the file (it will go into `.bss`), so perhaps that is the reason?

Is the code commit and answer I gave a few days ago working and sufficient for you given the constraints we talked about? If so I can go ahead and close the issue. Let me know 🙂
